test/test_s3.py

```diff
@@ -192,11 +203,110 @@ def test_put_object_file_object(self):
+                on_headers=self._on_request_headers,
+                on_progress=self._on_progress_cancel_after_first_chunk)
+            finished_future = self.s3_request.finished_future
+            try:
+                finished_future.result(self.timeout)
+            except Exception as e:
+                self.assertEqual(e.name, "AWS_ERROR_S3_CANCELED_SUCCESS")
+
+            # Result check
+            self.data_len = int(HttpHeaders(self.response_headers).get("Content-Length"))
+            self.assertLess(
+                self.transferred_len,
+                self.data_len,
+                "the cancel failed to block all the following body")
+
+            # The on_finish callback may invoke the progress
+            self.assertLessEqual(self.progress_invoked, 2)
+            shutdown_event = self.s3_request.shutdown_event
+            del self.s3_request
+            self.assertTrue(shutdown_event.wait(self.timeout))
+            os.remove(file.name)
+
+    def test_get_object_quick_cancel(self):
+        # a 5 GB file
+        request = self._get_object_request("/crt-canary-obj-single-part-9223372036854775807")
+        s3_client = s3_client_new(False, self.region, 5 * 1024 * 1024)
+        with NamedTemporaryFile(mode="w", delete=False) as file:
+            file.close()
+            s3_request = s3_client.make_request(
+                request=request,
+                recv_filepath=file.name,
+                type=S3RequestType.GET_OBJECT,
+                on_headers=self._on_request_headers,
+                on_progress=self._on_progress)
+            s3_request.cancel()
+            finished_future = s3_request.finished_future
+            try:
+                finished_future.result(self.timeout)
+            except Exception as e:
+                self.assertEqual(e.name, "AWS_ERROR_S3_CANCELED_SUCCESS")
+
+            # The on_finish callback may invoke the progress
+            shutdown_event = s3_request.shutdown_event
+            del s3_request
+            self.assertTrue(shutdown_event.wait(self.timeout))
+            os.remove(file.name)
+
+    def _put_object_cancel_helper(self, cancel_after_read):
+        read_futrue = Future()
+        put_body_stream = FakeReadStream(read_futrue)
+        data_len = 10 * 1024 * 1024 * 1024  # some fake length
+        headers = HttpHeaders([("host", self._build_endpoint_string(self.region, self.bucket_name)),
+                               ("Content-Type", "text/plain"), ("Content-Length", str(data_len))])
+        http_request = HttpRequest("PUT", "/cancelled_request", headers, put_body_stream)
+        s3_client = s3_client_new(False, self.region, 5 * 1024 * 1024)
+        s3_request = s3_client.make_request(
+            request=http_request,
+            type=S3RequestType.PUT_OBJECT,
+            on_headers=self._on_request_headers)
+
+        if cancel_after_read:
+            read_futrue.result(self.timeout)
+
+        s3_request.cancel()
```

Comment:
I don't know about python...
Here is a deadlock I met.
Since we will hold lock to read from inputstream, https://github.com/awslabs/aws-c-s3/blob/main/source/s3_auto_ranged_put.c#L300-L309, which means the on_progress callback is invoked with the lock hold. That's dangerous....
And not sure how it's not back to the C land and keep excute python code in the same thread and try to canel the meta request, which also will acquire a lock.
BOOM, dead lock.